### ISSUE TITLE
Error logged when adding Cloud Pub/Sub library

### ISSUE THREAD

**Issue by chanseokoh** [GoogleCloudPlatform/google-cloud-eclipse#2500]
~~Looking at the message, I think it has something to do with the `null` artifact qualifier.~~

```
Oct 16, 2017 6:26:19 PM com.google.cloud.tools.eclipse.appengine.libraries.model.Library resolveDependencies
SEVERE: Could not load library MavenCoordinates [repository=central, io.netty:netty-handler:jar:null:4.1.14.Final]
org.eclipse.core.runtime.CoreException: Could not resolve dependencies
	at com.google.cloud.tools.eclipse.util.DependencyResolver$1.call(DependencyResolver.java:87)
	at com.google.cloud.tools.eclipse.util.DependencyResolver$1.call(DependencyResolver.java:1)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.executeBare(MavenExecutionContext.java:176)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.execute(MavenExecutionContext.java:151)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.execute(MavenExecutionContext.java:99)
	at com.google.cloud.tools.eclipse.util.DependencyResolver.getTransitiveDependencies(DependencyResolver.java:95)
	at com.google.cloud.tools.eclipse.appengine.libraries.model.LibraryFactory.loadTransitiveDependencies(LibraryFactory.java:119)
	at com.google.cloud.tools.eclipse.appengine.libraries.model.Library.resolveDependencies(Library.java:207)
	at com.google.cloud.tools.eclipse.appengine.libraries.model.Library.getAllDependencies(Library.java:114)
	at com.google.cloud.tools.eclipse.appengine.libraries.repository.LibraryClasspathContainerResolverService.resolveLibraryFiles(LibraryClasspathContainerResolverService.java:187)
	at com.google.cloud.tools.eclipse.appengine.libraries.repository.LibraryClasspathContainerResolverService.resolveContainer(LibraryClasspathContainerResolverService.java:140)
	at com.google.cloud.tools.eclipse.appengine.libraries.LibraryClasspathContainerInitializer.initialize(LibraryClasspathContainerInitializer.java:84)
   ...
```

I queried the artifact on the Maven Central and noticed that all the io.netty:netty-handler releases have qualifiers (`Final`, `Alpha`, `Alpha2`, `Beta`, etc): https://search.maven.org/#search%7Cgav%7C1%7Cg%3A%22io.netty%22%20AND%20a%3A%22netty-handler%22
(BTW, `5.0.0.Alpah2` from 2015 has the highest major version, while the latest release is `4.1.16.Final` in Sep 2017.)

~~Perhaps this is due to https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/2488?~~

**Comment by chanseokoh** [comment]
No, actually, it is after #2489 that this started to happen.

**Comment by elharo** [comment]
This is a quick fix. Let's just allow "final" as a release qualifier.

**Comment by chanseokoh** [comment]
@elharo I said to early about #2488. This seems to happen after #2489.

**Comment by chanseokoh** [comment]
FYI, two more artifacts with the error:

`SEVERE: Could not load library MavenCoordinates [repository=central, io.grpc:grpc-auth:jar:null:1.7.0]`

`SEVERE: Could not load library MavenCoordinates [repository=central, io.grpc:grpc-netty:jar:null:1.7.0]`

**Comment by elharo** [comment]
#2501 should fix them both

**Comment by chanseokoh** [comment]
Sorry, I hastily guessed that this has something to do with qualifiers. Turns out, it doesn't likely. The stack trace doesn't go through `ArtifactRetriever` either. I think something happened with #2489, not #2488.

I also need to clarify the issue. There is no visual error to users; it just logs the errors. Interestingly, even if the errors are logged for the three artifacts, in the end, I see them in the master container.

**Comment by elharo** [comment]
Interesting. As long as they show up in master container I'm marking this low priority.

**Comment by briandealwis** [comment]
Ah, our exception is incomplete: `DependencyResolver#getTransitiveDependencies()` is [swallowing the resolver exceptions](https://github.com/GoogleCloudPlatform/google-cloud-eclipse/blob/master/plugins/com.google.cloud.tools.eclipse.util/src/com/google/cloud/tools/eclipse/util/DependencyResolver.java#L87).

**Comment by chanseokoh** [comment]
#2503 surfaces root causes. I see two different errors.

1. `No versions available for io.grpc:grpc-core:jar:[1.7.0] within specified range`

   ```
   Caused by: org.eclipse.aether.collection.DependencyCollectionException: Failed to collect dependencies at io.grpc:grpc-auth:jar:1.7.0 -> io.grpc:grpc-core:jar:[1.7.0]
   	at org.eclipse.aether.internal.impl.DefaultDependencyCollector.collectDependencies(DefaultDependencyCollector.java:291)
   	at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:341)
   	... 75 more
   Caused by: org.eclipse.aether.resolution.VersionRangeResolutionException: No versions available for io.grpc:grpc-core:jar:[1.7.0] within specified range
   	at org.eclipse.aether.internal.impl.DefaultDependencyCollector.filterVersions(DefaultDependencyCollector.java:648)
   ```

   ```
   Caused by: org.eclipse.aether.collection.DependencyCollectionException: Failed to collect dependencies at io.grpc:grpc-netty:jar:1.7.0 -> io.grpc:grpc-core:jar:[1.7.0]
   	at org.eclipse.aether.internal.impl.DefaultDependencyCollector.collectDependencies(DefaultDependencyCollector.java:291)
   	at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:341)
   	... 75 more
   Caused by: org.eclipse.aether.resolution.VersionRangeResolutionException: No versions available for io.grpc:grpc-core:jar:[1.7.0] within specified range
   	at org.eclipse.aether.internal.impl.DefaultDependencyCollector.filterVersions(DefaultDependencyCollector.java:648)
   ```

1. `Illegal character in path at index 101: https://repo.maven.apache.org/maven2/io/netty/netty-tcnative/2.0.5.Final/netty-tcnative-2.0.5.Final-${os.detected.classifier}.jar`

   ```
   Caused by: org.eclipse.aether.resolution.ArtifactResolutionException: Failure to transfer io.netty:netty-tcnative:jar:${os.detected.classifier}:2.0.5.Final from https://repo.maven.apache.org/maven2/ was cached in the local repository, resolution will not be reattempted until the update interval of central has elapsed or updates are forced. Original error: Could not transfer artifact io.netty:netty-tcnative:jar:${os.detected.classifier}:2.0.5.Final from/to central (https://repo.maven.apache.org/maven2/): Illegal character in path at index 101: https://repo.maven.apache.org/maven2/io/netty/netty-tcnative/2.0.5.Final/netty-tcnative-2.0.5.Final-${os.detected.classifier}.jar
   	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:444)
   	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:246)
   	at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:367)
   	... 75 more
   Caused by: org.eclipse.aether.transfer.ArtifactTransferException: Failure to transfer io.netty:netty-tcnative:jar:${os.detected.classifier}:2.0.5.Final from https://repo.maven.apache.org/maven2/ was cached in the local repository, resolution will not be reattempted until the update interval of central has elapsed or updates are forced. Original error: Could not transfer artifact io.netty:netty-tcnative:jar:${os.detected.classifier}:2.0.5.Final from/to central (https://repo.maven.apache.org/maven2/): Illegal character in path at index 101: https://repo.maven.apache.org/maven2/io/netty/netty-tcnative/2.0.5.Final/netty-tcnative-2.0.5.Final-${os.detected.classifier}.jar
   	at org.eclipse.aether.internal.impl.DefaultUpdateCheckManager.newException(DefaultUpdateCheckManager.java:238)
   ```

**Comment by chanseokoh** [comment]
I was testing #2502, and #2502 seems to fix this too! Without #2502, we were updating transitive deps, so we try to get grpc-core 1.7.0 for example. After #2502, we correctly get grpc-core 1.6.1, and the problem goes away.
